File: src/analysis_lib/examples/analyze_example.py
#!/usr/bin/env python3
"""
Example script demonstrating the Mixxx Analysis Library Python wrapper

Usage:
    python analyze_example.py /path/to/audio/file.mp3
    python analyze_example.py /music/folder/*.mp3
"""

import sys
from pathlib import Path

# Add parent directory to path to import the wrapper
sys.path.insert(0, str(Path(__file__).parent.parent))

from mixxx_analysis import MixxxAnalyzer


def analyze_single_file(analyzer, file_path):
    """Analyze a single audio file and print results."""
    print(f"\nAnalyzing: {file_path}")
    print("-" * 60)
    
    try:
        result = analyzer.analyze_file(str(file_path))
        
        # Print BPM
        if result['bpm_detected']:
            print(f"  BPM:          {result['bpm']:.2f} ✓")
        else:
            print(f"  BPM:          Not detected ✗")
        
        # Print Key
        if result['key_detected']:
            print(f"  Key:          {result['key_name']} (value: {result['key']}) ✓")
        else:
            print(f"  Key:          Not detected ✗")
        
        # Print ReplayGain
        if result['gain_calculated']:
            print(f"  ReplayGain:   {result['replay_gain_db']:.2f} dB (ratio: {result['replay_gain_ratio']:.4f}) ✓")
        else:
            print(f"  ReplayGain:   Not calculated ✗")
        
        if result['error_message']:
            print(f"  Warning:      {result['error_message']}")
        
        return result
        
    except FileNotFoundError:
        print(f"  ERROR: File not found")
        return None
    except Exception as e:
        print(f"  ERROR: {e}")
        return None


def analyze_batch(file_paths):
    """Analyze multiple audio files."""
    results = []
    
    with MixxxAnalyzer() as analyzer:
        print(f"Mixxx Analysis Library v{analyzer.get_version()}")
        print(f"Analyzing {len(file_paths)} file(s)...\n")
        
        for file_path in file_paths:
            result = analyze_single_file(analyzer, file_path)
            if result:
                results.append((file_path, result))
    
    return results


def main():
    if len(sys.argv) < 2:
        print("Usage: python analyze_example.py <audio_file> [<audio_file2> ...]")
        print("\nExample:")
        print("  python analyze_example.py song.mp3")
        print("  python analyze_example.py *.mp3")
        print("  python analyze_example.py /music/folder/*.flac")
        sys.exit(1)
    
    # Collect all file paths
    file_paths = []
    for arg in sys.argv[1:]:
        path = Path(arg)
        if path.exists():
            if path.is_file():
                file_paths.append(path)
            elif path.is_dir():
                # If directory, find all audio files
                for ext in ['*.mp3', '*.flac', '*.ogg', '*.wav', '*.m4a']:
                    file_paths.extend(path.glob(ext))
    
    if not file_paths:
        print("ERROR: No valid audio files found")
        sys.exit(1)
    
    # Analyze all files
    results = analyze_batch(file_paths)
    
    # Print summary
    print("\n" + "=" * 60)
    print(f"Summary: Analyzed {len(results)}/{len(file_paths)} file(s) successfully")
    
    if results and len(results) > 1:
        print("\nQuick Summary:")
        for file_path, result in results:
            status = []
            if result['bpm_detected']:
                status.append(f"BPM:{result['bpm']:.1f}")
            if result['key_detected']:
                status.append(f"Key:{result['key_name']}")
            print(f"  {file_path.name}: {', '.join(status) if status else 'No results'}")


if __name__ == "__main__":
    main()

File: src/analysis_lib/examples/analyze_example.c
/**
 * Example program demonstrating the Mixxx Analysis Library C API
 * 
 * Compile with:
 *   gcc -o analyze_example analyze_example.c -lmixxx_analysis
 * 
 * Run with:
 *   ./analyze_example /path/to/audio/file.mp3
 */

#include "mixxx_analysis.h"
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <audio_file>\n", argv[0]);
        return 1;
    }
    
    const char* file_path = argv[1];
    
    /* Print library version */
    printf("Mixxx Analysis Library v%s\n", mixxx_analyzer_version());
    printf("Analyzing: %s\n\n", file_path);
    
    /* Create analyzer context */
    MixxxAnalyzerHandle analyzer = mixxx_analyzer_create();
    if (!analyzer) {
        fprintf(stderr, "ERROR: Failed to create analyzer context\n");
        return 1;
    }
    
    /* Analyze the file */
    MixxxAnalysisResult result;
    int status = mixxx_analyze_file(analyzer, file_path, &result);
    
    if (status != 0) {
        fprintf(stderr, "ERROR: Analysis failed: %s\n", result.error_message);
        mixxx_analyzer_destroy(analyzer);
        return 1;
    }
    
    /* Print results */
    printf("Analysis Results:\n");
    printf("-----------------\n");
    
    if (result.bpm_detected) {
        printf("BPM:          %.2f\n", result.bpm);
    } else {
        printf("BPM:          Not detected\n");
    }
    
    if (result.key_detected) {
        printf("Key:          %s (value: %d)\n", result.key_name, result.key);
    } else {
        printf("Key:          Not detected\n");
    }
    
    if (result.gain_calculated) {
        printf("ReplayGain:   %.2f dB (ratio: %.4f)\n", 
               result.replay_gain_db, result.replay_gain_ratio);
    } else {
        printf("ReplayGain:   Not calculated\n");
    }
    
    /* Cleanup */
    mixxx_analyzer_destroy(analyzer);
    
    printf("\nAnalysis complete!\n");
    return 0;
}
